src/retriever/utils.ts

```diff
@@ -0,0 +1,129 @@
+/**
+ * Utility to attempt fetching a piece from multiple providers in parallel.
+ */
+
+import type { ApprovedProviderInfo, CommP } from '../types.js'
+import { constructPieceUrl, constructFindPieceUrl } from '../utils/piece.js'
+import { createError } from '../utils/errors.js'
+
+// Define the type for provider attempt results (internal to this function)
+interface ProviderAttemptResult {
+  response: Response
+  index: number
+}
+
+/**
+ * Attempt to fetch a piece from multiple providers in parallel
+ * @param providers - List of providers to try
+ * @param commp - The piece to fetch
+ * @param retrieverName - Name of the calling retriever for error reporting
+ * @param signal - Optional abort signal
+ * @returns The first successful response
+ */
+export async function fetchPiecesFromProviders (
+  providers: ApprovedProviderInfo[],
```

Comment:
let's add a dedupe for `providers` inside this function, as I mentioned in Slack, it's possible to get the same provider more than once in here if oyu've stored the same piece with them more than once

Reply:
actually, maybe we should do it up in the subgraph retriever, if you look in chain.ts you'll see that we already do it there

if we can't do a uniqueness check in the graphql query then let's do it when we get the response